vdk-meta-jobs: Meta Jobs DAG validation

yonitoo opened the pull request:
Introduce a DagValidator class that properly validates the Meta Jobs DAG before it is built. It checks for:
 * DAG structure (spelling mistakes, all the types of the job dict keys)
 * duplicate jobs
 * DAG cycles

Rework the test_meta_job tests by introducing a TestMetaJob class that contains all the existing ones as well as some new validation-related ones.

Tested-by: all the existing tests pass and introduced some new unit tests for the different cases

Signed-off-by: Yoan Salambashev <ysalambashev@vmware.com>

gageorgiev commented on projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/meta_dag.py:
```diff
@@ -134,3 +136,110 @@ def _start_job(self, node):
+                    "",
+                    "Meta Job failed due to a data job validation failure.",
+                    f"Validation of Data Job failed. "
+                    "Some keys in the Data Job Dict are not allowed or some of the required ones are missing.",
+                    "The DAG will not be built and the Meta Job will fail.",
+                    "Check the Data Job Dict keys for errors and make sure the required ones are present.",
```
It would be good to list which keys are allowed and which are required to make debugging easier.

yonitoo replied:
Makes sense, divided them into two conditions. Will push soon.

gageorgiev commented on projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/meta_dag.py:
```diff
@@ -134,3 +136,110 @@ def _start_job(self, node):
+        if any(key not in allowed_job_keys for key in job.keys()) or any(
+            key not in job for key in required_job_keys
+        ):
+            raise UserCodeError(
+                ErrorMessage(
```
Do we need to raise an error in the case where extra keys are included in the job dict? Perhaps a warning is sufficient. Maybe @tozka could comment.

gageorgiev commented on projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/meta_dag.py:
```diff
@@ -134,3 +136,110 @@ def _start_job(self, node):
+                    f"Check the Data Job Dict value of job_name. Expected type is string.",
+                )
+            )
+        if job["job_name"] in registered_job_names:
+            raise UserCodeError(
+                ErrorMessage(
+                    "",
+                    "Meta Job failed due to a data job validation failure.",
+                    f"Validation of Data Job {job['job_name']} failed. "
+                    f"Job with such name already exists.",
+                    "The DAG will not be built and the Meta Job will fail.",
+                    "Check the Data Job Dict value of job_name.",
+                )
+            )
```
Can you explain this fail condition? I don't think I understand it.

yonitoo replied:
Changed it a bit, must be more comprehensible now. The idea of this condition is to check if a job is already validated (exists).

gageorgiev commented on projects/vdk-plugins/vdk-meta-jobs/tests/test_meta_job.py:
```diff
@@ -305,6 +305,111 @@ def test_meta_job_circular_dependency(httpserver: PluginHTTPServer):
 
 
+def test_meta_job_wrong_job_type(httpserver: PluginHTTPServer):
+    jobs = [("job" + str(i), [200], "succeeded") for i in range(1, 5)]
+    api_url = _prepare(httpserver, jobs)
+
+    with mock.patch.dict(
+        os.environ,
+        {"VDK_CONTROL_SERVICE_REST_API_URL": api_url},
+    ):
```
You can avoid repeating this chunk so many times is you wrap these tests in a class and including it in a setUp method; read more here https://docs.python.org/3/library/unittest.html#unittest.TestCase.setUp

gageorgiev reviewed (approved):
looks good overall, I'm approving it but I'd still appreciate responses for my comments

tozka commented on projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/meta_dag.py:
```diff
@@ -134,3 +136,110 @@ def _start_job(self, node):
             else:
                 raise
+
+    def _validate_job(self, job: Dict):
```
Also consider moving the logic in separate class: 

 This will follow better the spirit  SRP([Single responsibility principle](https://en.wikipedia.org/wiki/Single-responsibility_principle)) and better separation of concerns
 
 DagValidator with public method validate(jobs: List[Dict]) ;

yonitoo replied:
Created the suggested new class and moved the logic there.

tozka reviewed (commented):
In addition to checking for job structure, there are a few other validations you may want to consider on DAG level . (if you are planning to add them in separate PR , then you can ignore this comment) 

1. Validate that there are not cycles  
2. As the DAG is a list of jobs it's possible to have duplicate job names.  This can lead to confusion and unintended behavior 

You can use graphlib.TopologicalSorter for that (at least for 1.)

yonitoo commented on projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/meta_dag.py:
```diff
@@ -134,3 +136,110 @@ def _start_job(self, node):
+                ErrorMessage(
+                    "",
+                    "Meta Job failed due to a data job validation failure.",
+                    f"Validation of Data Job failed. "
```
Makes sense, yeah. Removed the redundant part.

yonitoo commented:
Added 2 new classes:
-> DagValidator: the validation logic is moved there
-> TestMetaJob: cleaned up the mess around test_meta_job unit tests 

The only thing left to be done now is to add topological sorter validation.

tozka commented on projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/dag_validator.py:
```diff
@@ -0,0 +1,127 @@
+required_job_keys = ["job_name", "depends_on"]
+
+
+class DagValidator:
```
Please write documentation about the purpose and usage of the calss.

tozka commented on projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/dag_validator.py:
```diff
@@ -0,0 +1,127 @@
+
+
+class DagValidator:
+    def validate(self, jobs: List[Dict]):
```
All public functions/method shoud have documentation explain their purpose and usage.

yonitoo replied:
Added docs for the public validate function.

tozka commented on projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/dag_validator.py:
```diff
@@ -0,0 +1,127 @@
+        self._validate_team_name(job)
+        self._validate_fail_meta_job_on_error(job)
+        log.info(f"Successfully validated job: {job['job_name']}")
+        validated_jobs.append(job["job_name"])
+        return validated_jobs
```
Why here? 

The _validate_job method should only be responsible for validating a single job, and not for keeping track of the list of validated jobs. (See [single responsibility principle](https://en.wikipedia.org/wiki/Single-responsibility_principle))

I think it would be cleaner and more explicit to add new method `self._validate_no_duplicates(jobs)` currently this validation is hidden behind _validate_job_name which again breaks SRP as it does more than one thing which I initially missed and found validated_jobs very strange.

yonitoo replied:
I dunno why I added this logic here. It totally breaks the SRP, changed it.

tozka commented on projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/dag_validator.py:
```diff
@@ -0,0 +1,127 @@
+        validated_jobs.append(job["job_name"])
+        return validated_jobs
+
+    def _validate_allowed_and_required_keys(self, job):
```
add type hints

yonitoo replied:
Done.

tozka commented on projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/dag_validator.py:
```diff
@@ -0,0 +1,127 @@
+                f"are missing. Required keys: {required_job_keys}.",
+            )
+
+    def _validate_job_name(self, job, validated_jobs):
```
add type hints (here and everywhere)

yonitoo replied:
Done.

tozka commented on projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/dag_validator.py:
```diff
@@ -0,0 +1,127 @@
+        for job in jobs:
+            validated_jobs = self._validate_job(job, validated_jobs)
+
+    def _raise_error(self, job, error_type, countermeasures):
```
type hints

yonitoo replied:
Done.

tozka commented on projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/dag_validator.py:
```diff
@@ -0,0 +1,127 @@
+ERROR = Error(
+    TYPE="type", PERMISSION="permission", REQUIREMENT="requirement", CONFLICT="conflict"
+)
+allowed_job_keys = ["job_name", "team_name", "fail_meta_job_on_error", "depends_on"]
```
use set instead of list
```suggestion
allowed_job_keys = {"job_name", "team_name", "fail_meta_job_on_error", "depends_on"}
```

yonitoo replied:
Done.

tozka commented on projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/dag_validator.py:
```diff
@@ -0,0 +1,127 @@
+        return validated_jobs
+
+    def _validate_allowed_and_required_keys(self, job):
+        if any(key not in allowed_job_keys for key in job.keys()):
+            forbidden_keys = [key not in allowed_job_keys for key in job.keys()]
```
You can streamline the code (and remove minor duplication)
```
     forbidden_keys = [key not in allowed_job_keys for key in job.keys()]
     if forbidden_keys: 
           self._raise_error
     missing_keys = [key not in job for key in required_job_keys]
     if missing_keys:  
        ...

yonitoo replied:
Done.

tozka reviewed (commented):
Functionality looks pretty good. I have mostly clean code suggestion.

tozka reviewed (approved):
LGTM. 

Please go over the message description. Make sure they are clear to someone without context . They know what the problem is and how to solve it.
